blockchain/indexers/common.go

```diff
@@ -173,3 +182,909 @@ func interruptRequested(ctx context.Context) bool {
+	// complete.
+	err = markIndexDeletion(db, idxKey)
+	if err != nil {
+		return err
+	}
+
+	// Remove the index tip, version, bucket, and in-progress drop flag.
+	// Removing the index bucket also recursively removes all values saved to
+	// the index.
+	err = dropIndexMetadata(db, idxKey, idxName)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Dropped %s", idxName)
+	return nil
+}
+
+// markIndexDeletion marks the index identified by idxKey for deletion.  Marking
+// an index for deletion allows deletion to resume next startup if an
+// incremental deletion was interrupted.
+func markIndexDeletion(db database.DB, idxKey []byte) error {
+	return db.Update(func(dbTx database.Tx) error {
+		indexesBucket := dbTx.Metadata().Bucket(indexTipsBucketName)
+		return indexesBucket.Put(indexDropKey(idxKey), idxKey)
+	})
+}
+
+// tip returns the current tip hash and height of the provided index.
+func tip(db database.DB, key []byte) (int64, *chainhash.Hash, error) {
+	var hash *chainhash.Hash
+	var height int32
+	err := db.View(func(dbTx database.Tx) error {
+		var err error
+		hash, height, err = dbFetchIndexerTip(dbTx, key)
+		return err
+	})
+	if err != nil {
+		return 0, nil, err
+	}
+	return int64(height), hash, err
+}
+
+// recover reverts the index to a block on the main chain by repeatedly
+// disconnecting the index tip if it is not on the main chain.
+func recover(ctx context.Context, idx Indexer) error {
+	// Fetch the current tip for the index.
+	height, hash, err := idx.Tip()
+	if err != nil {
+		return err
+	}
+
+	// Nothing to do if the index does not have any entries yet.
+	if height == 0 {
+		return nil
+	}
+
+	// Nothing to do if the index tip is on the main chain.
+	if idx.Queryer().MainChainHasBlock(hash) {
+		return nil
+	}
+
+	log.Infof("%s: recovering from tip %d (%s)", idx.Name(), height, hash)
+
+	// Create a progress logger for the recovery process below.
+	progressLogger := progresslog.NewBlockProgressLogger("Recovered", log)
+
+	queryer := idx.Queryer()
+	var cachedBlock *dcrutil.Block
+	for !queryer.MainChainHasBlock(hash) {
+		if interruptRequested(ctx) {
+			return errInterruptRequested
+		}
+
+		// Get the block, unless it's already cached.
+		var block *dcrutil.Block
+		if cachedBlock == nil && height > 0 {
+			block, err = queryer.BlockByHash(hash)
+			if err != nil {
+				return err
+			}
+		} else {
+			block = cachedBlock
+		}
+
+		parentHash := &block.MsgBlock().Header.PrevBlock
+		parent, err := queryer.BlockByHash(parentHash)
+		if err != nil {
+			return err
+		}
+		cachedBlock = parent
+
+		var prevScripts PrevScripter
+		err = idx.DB().Update(func(dbTx database.Tx) error {
+			if interruptRequested(ctx) {
+				return errInterruptRequested
+			}
+
+			// Fetch the associated script information for previous outputs
+			// of the block if the index requires them.
+			if indexNeedsInputs(idx) {
+				var err error
+				prevScripts, err = queryer.PrevScripts(dbTx, block)
```

Comment:
There is one issue with the recovery code.  This panics when attempting to fetch the spend journal because it will no longer exist since the spend journal is removed when disconnecting blocks.

You can simulate this easily by:

- Enabling indexes and start syncing
- Restart `dcrd` with indexes disabled
- Use `invalidateblock` to invalidate a block at a previous height
- Restart `dcrd` with indexes enabled
- The index will need to rollback to the block before the invalidated block (where the tip currently is).  It will crash since the spend journal is not available for those blocks that it needs to roll back.

Reply:
After going back and forth on how best to resolve this issue I think the best way forward would be to track spend journal entry removals and only remove them once they are out of range for a reorg per the max reorg depth. This would also mean blocks connected to the chain will have to ensure their associated spend journals are not scheduled to be removed, if they are the action is cancelled since the block has been reconnected back to the chain. This tracking would need to be persisted to db. If there are no objections or suggestions of a better solution I'll proceed to implement this in a different PR.

Reply:
Can you summarize the solution in the current PR around this?  This comment made it sound like you wanted to defer fixing this, but I see a bunch of logic for a `SpendJournalFilterer` was added.  So I just want to make sure I understand the approach before I continue reviewing.

Reply:
Sure, the solution was to defer the removal of the spend journal entries for the blocks being disconnected since its possible they will be needed later in a recovery process. For block invalidation, all block hashes being disconnected are added to the prune filter to prevent their removal. The prune filter is updated on startup particularly for indexes that need inputs from transactions. The filter keeps count of the number of processes that require the spend journal associated with the block hash by incrementing the count when `AddEntry` is called, `RemoveEntry` decrements the count until there are no processes needing it, at that point the spend journal entry is finally deleted.